[lldb-dap] Adding support for cancelling a request.

ashgti opened the pull request:
Adding support for cancelling requests.

There are two forms of request cancellation.

* Preemptively cancelling a request that is in the queue.
* Actively cancelling the in progress request as a best effort attempt using `SBDebugger.RequestInterrupt()`.

vogelsgesang commented on lldb/tools/lldb-dap/Handler/CancelRequestHandler.cpp:
```diff
@@ -0,0 +1,56 @@
+// },
+llvm::Expected<CancelResponseBody>
+CancelRequestHandler::Run(const CancelArguments &arguments) const {
+  /* no-op, simple ack of the request. */
```
seems weird... I would have expected that we do something here. Or at least a comment which explains me, why we not don't have to do anything here...

This comment currently explains the "what?" (which is kind of obvious from the code). I would prefer a comment explaining the "why?" (which is much harder to reconstruct by reading the code)

ashgti replied:
I updated the comment to explain that this is handled in the DAP::Loop instead.

vogelsgesang commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -256,7 +257,27 @@ void DAP::SendJSON(const llvm::json::Value &json) {
+  lldb_private::Status status;
+  // If the debugger was interrupted while handling a response, mark the
+  // response as cancelled since it may contain partial information.
+  if (const auto *resp = std::get_if<protocol::Response>(&M);
+      debugger.InterruptRequested()) {
```
```suggestion
  if (const auto *resp = std::get_if<protocol::Response>(&M);
      resp && debugger.InterruptRequested()) {
```

ashgti replied:
Done.

vogelsgesang reviewed (commented):
just a very cursory review. I will review this more closely as soon as the preparatory commits landed. Those stacked PRs are a bit hard to review

vogelsgesang commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -857,29 +819,112 @@ lldb::SBError DAP::Disconnect(bool terminateDebuggee) {
+  std::condition_variable queue_cv;
+  std::mutex queue_mutex;
+  std::thread queue_reader([&]() {
+    auto cleanup = llvm::make_scope_exit([&]() {
```
```suggestion
    auto cleanup = llvm::make_scope_exit([&]() {
      llvm::set_thread_name(dap.client_name + ".transport_handler");
```

ashgti replied:
Done.

ashgti commented on lldb/tools/lldb-dap/Protocol.h:
```diff
@@ -348,6 +351,71 @@ llvm::json::Value toJSON(const ExitedEventBody &);
+//     acknowledgement, so no body field is required."
+//   }]
+// },
+using CancelResponseBody = VoidResponseBody;
```
This was handled in an upstream patch.

ashgti commented on lldb/tools/lldb-dap/Protocol.h:
```diff
@@ -152,6 +153,8 @@ struct Response {
 bool fromJSON(const llvm::json::Value &, Response &, llvm::json::Path);
 llvm::json::Value toJSON(const Response &);
 
+using VoidResponseBody = std::nullptr_t;
```
Done.

ashgti commented on lldb/tools/lldb-dap/Transport.cpp:
```diff
@@ -67,7 +67,7 @@ static Status ReadExpected(IOObjectSP &descriptor, llvm::StringRef expected) {
   if (expected != result) {
-    return Status::FromErrorStringWithFormatv("expected %s, got %s", expected,
+    return Status::FromErrorStringWithFormatv("expected {0}, got {1}", expected,
                                               result);
```
Done.

ashgti committed:
[lldb-dap] Adding support for cancelling a request.

Adding support for cancelling requests.

There are two forms of request cancellation.

* Preemptively cancelling a request that is in the queue.
* Actively cancelling the in progress request as a best effort attempt using `SBDebugger.RequestInterrupt()`.

vogelsgesang commented on lldb/test/API/tools/lldb-dap/launch/TestDAP_launch.py:
```diff
@@ -27,6 +27,7 @@ def test_default(self):
         lines = output.splitlines()
         self.assertIn(program, lines[0], "make sure program path is in first argument")
 
+    @skipIfWindows
```
why? Did this work previously and was now broken by cancellation support?

vogelsgesang commented on lldb/tools/lldb-dap/Handler/InitializeRequestHandler.cpp:
```diff
@@ -463,6 +463,8 @@ void InitializeRequestHandler::operator()(
   body.try_emplace("supportsDataBreakpoints", true);
   // The debug adapter supports the `readMemory` request.
   body.try_emplace("supportsReadMemoryRequest", true);
+  // The debug adapter supports the `cancel` request.
+  body.try_emplace("supportsCancelRequest", true);
```
move to `CancelRequestHandler` (#131943)

vogelsgesang commented on lldb/tools/lldb-dap/Transport.cpp:
```diff
@@ -96,7 +115,7 @@ Expected<std::optional<Message>> Transport::Read() {
     return createStringError(
         formatv("invalid content length {0}", *raw_length).str());
 
-  Expected<std::string> raw_json = ReadFull(*input, length);
```
we should probably only apply a timeout before receiving the first byte of a message.

Otherwise, we might run into hard-to-debug issues where the client sends

```
Content-Length: 123
\r\n\r\n
<wait for 2 seconds>
actual request body
```

With the current logic, we would first consume the `Content-Length: 123\r\n\rn\n` header, then run into the timeout. Upon retrying the read in `DAP::Loop()` we would the find the request body without the required `Content-Length` header.

The client would be compliant with the Debug Adapter Protocol specification, yet `lldb-dap` would choke on this message.

It seems we are only using the timeout such that the `disconnecting` flag is checked regularly in `DAP::Loop`. Instead of using a timeout to wake up the reader-thread, would it maybe make sense to instead call `Transport::Close` when we want to shut down the reader? That should also cancel any outstanding reads, doesn't it?

ashgti replied:
> we should probably only apply a timeout before receiving the first byte of a message.
> 
> Otherwise, we might run into hard-to-debug issues where the client sends

Yea, thats a good idea, I'll only apply the timeout to the message header.

> It seems we are only using the timeout such that the `disconnecting` flag is checked regularly in `DAP::Loop`. Instead of using a timeout to wake up the reader-thread, would it maybe make sense to instead call `Transport::Close` when we want to shut down the reader? That should also cancel any outstanding reads, doesn't it?

Closing a file descriptor does not interrupt in progress reads, I think its platform dependent on what happens.

ashgti replied:
Also, frustrating on Windows stdin is an anonymous pipe and anonymous pipes do not support async reads (https://learn.microsoft.com/en-us/windows/win32/ipc/anonymous-pipe-operations) which is why I disabled the one test on Windows, since we can't read with a timeout on stdin.

labath replied:
> Closing a file descriptor does not interrupt in progress reads, I think its platform dependent on what happens.

Very much so. It's also basically impossible to do this in race-free manner,.

> Also, frustrating on Windows stdin is an anonymous pipe and anonymous pipes do not support

I don't know if this helps you (since you still have to use a dedicated API instead of `select(2)`), but `lldb_private::Pipe` works around this by creating a named pipe with an random name (from what I've read, that's actually how windows implements anonymous pipes internally).

ashgti committed:
Only apply the Transport::Read timeout to the initial read of the header, once we have the header we should read the full message.

ashgti committed:
Fixing missing includes and rebasing on main.

ashgti committed:
Applying clang-format.

JDevlieghere commented:
I left a few small comments but overall this looks good. 

The approach is slightly different from how I expected you to implement this. Note that I don't know if it's actually better, it's more of a "if I had to implement this, that's how I would have started." You've been working on this for a bit, so I'm trusting your judgement. 

Anyway, what I had in mind when I read your RFC and was doing the RequestHandler work, triggered by a question from @labath, is that instead of having one RequestHandler instance per command (which is what we have today), we could instantiate a RequestHandler per incoming command (or a new object that has a reference to the request hander and stores the arguments). 

The read thread could decode its arguments and store away the protocol in the RequestHandler/new object and add it to the queue. Beyond that, things work pretty much how you've implemented them here, but with the logic that checks if the request has been cancelled in the RequestHandler. More concretely the operator doesn't take arguments anymore, and it's implemented something like this:

```
void operator() {
  if (IsCancelled()) // Check if our request ID is in the cancelled request set.
    RespondCancelled();
  
  llvm::Expected<Body> body = Run(); 
  if (!body) 
    // Handle error

  if (IsInterrupted()) // Check if the SBDebugger was interrupted. 
  	RespondCancelled(); 
}
```

If we go with the new object (`PendingRequest`?) the request handlers can remain what they are today, or if we want to store the data in them we'll need to find a way to dispatch to them (similar to my comment in yesterday's PR). 

Anyway, like I said, not sure if this is actually better, just sharing my thoughts.

labath reviewed (commented):
I think there's a race in the cancellation handling:
1. request handling thread sets active_seq (but doesn't start executing the request yet)
2. cancellation request comes in, queue_reader sets the interrupt flag
3. request handling thread begins handling the request, starts by clearing the flag
4. request is not cancelled

I think this would be easier to guarantee correctness (and review) if `active_seq` was *not* an atomic and its modification, as well as the acts of setting and clearing the interrupt flag were done under a mutex

JDevlieghere reviewed (commented):
(looks like I didn't submit my comments yesterday)

ashgti commented:
> Anyway, what I had in mind when I read your RFC and was doing the RequestHandler work, triggered by a question from @labath, is that instead of having one RequestHandler instance per command (which is what we have today), we could instantiate a RequestHandler per incoming command (or a new object that has a reference to the request hander and stores the arguments).
> 
> The read thread could decode its arguments and store away the protocol in the RequestHandler/new object and add it to the queue. Beyond that, things work pretty much how you've implemented them here, but with the logic that checks if the request has been cancelled in the RequestHandler. More concretely the operator doesn't take arguments anymore, and it's implemented something like this:
> 
> ```
> void operator() {
>   if (IsCancelled()) // Check if our request ID is in the cancelled request set.
>     RespondCancelled();
>   
>   llvm::Expected<Body> body = Run(); 
>   if (!body) 
>     // Handle error
> 
>   if (IsInterrupted()) // Check if the SBDebugger was interrupted. 
>   	RespondCancelled(); 
> }
> ```
> 
> If we go with the new object (`PendingRequest`?) the request handlers can remain what they are today, or if we want to store the data in them we'll need to find a way to dispatch to them (similar to my comment in yesterday's PR).
> 
> Anyway, like I said, not sure if this is actually better, just sharing my thoughts.

I do think this is an interesting approach. I can take a look at this as trying to wrap some of this into some helpers or maybe making the RequestHandler allocate a new instance per request.

My logic behind the `lldb_dap::protocol` types was to keep them as POD data and to not have much (if any) logic in them so that if we need to move the data between threads or queues we can be relatively confident that we're not dealing data races. In swift terms (I'm more familiar with swift than c++), the `lldb_dap::protocol` types would be `struct`'s that are all `Sendable`.

That may not be as idiomatic in c++ though.

I do like the more simplified `operator()` idea to help deal with the cancelled / interrupted, my existing solution was done to support the LegacyRequestHandlers since they're sending the responses as raw `json::Value`'s still, but with the more unified approach we could handle the cancellation in the `operator()` body.

I'll try to see if I can come up with a `PendingRequest` or maybe take a look at allocating the `RequestHandler`'s in the queue.

JDevlieghere commented:
> My logic behind the `lldb_dap::protocol` types was to keep them as POD data and to not have much (if any) logic in them so that if we need to move the data between threads or queues we can be relatively confident that we're not dealing data races. In swift terms (I'm more familiar with swift than c++), the `lldb_dap::protocol` types would be `struct`'s that are all `Sendable`.

Totally, and I think that's a great design, and not something that would change in my suggestion, right? 

> I do like the more simplified `operator()` idea to help deal with the cancelled / interrupted, my existing solution was done to support the LegacyRequestHandlers since they're sending the responses as raw `json::Value`'s still, but with the more unified approach we could handle the cancellation in the `operator()` body.

If we want to support cancellation before moving all the request over to the protocol class, can we *temporarily* store both the JSON and the Protocol in the class we instantiate per request (whether that's the RequestHandler itself, or a class wrapping it) and use whichever one we need for that given request?

ashgti commented:
> If we want to support cancellation before moving all the request over to the protocol class, can we _temporarily_ store both the JSON and the Protocol in the class we instantiate per request (whether that's the RequestHandler itself, or a class wrapping it) and use whichever one we need for that given request?

For the `LegacyRequestHandler`, its more to do with the response than the request. We can store the request as a `protocol::Message` and in the `LegacyRequestHandler` we convert it back into a `json::Value` before we pass it to the handler. But the `LegacyRequestHandler`'s all tend to call `dap.SendJSON` directly with a `json::Value`. In my current impl of the cancel request, I check for the cancelled requests prior to sending them https://github.com/llvm/llvm-project/blob/1c2d19b901390c6080c1367341115150661a6a14/lldb/tools/lldb-dap/DAP.cpp#L245-L258

ashgti committed:
Adding fine grained locks for the DAP processing queue and tried to simplify the Loop reader.

ashgti committed:
Ensure we clear the cancel requests that have been processed.

ashgti commented:
> I think there's a race in the cancellation handling:
> 
> 1. request handling thread sets active_seq (but doesn't start executing the request yet)
> 2. cancellation request comes in, queue_reader sets the interrupt flag
> 3. request handling thread begins handling the request, starts by clearing the flag
> 4. request is not cancelled
> 
> I think this would be easier to guarantee correctness (and review) if `active_seq` was _not_ an atomic and its modification, as well as the acts of setting and clearing the interrupt flag were done under a mutex

Switched to a mutex for tracking the active request.

ashgti committed:
Using a few helpers to simplify the cancel responses.

ashgti commented on lldb/tools/lldb-dap/Handler/CancelRequestHandler.cpp:
```diff
@@ -0,0 +1,55 @@
+/// specific request issued earlier
+/// - to cancel a progress sequence.
+///
+/// Clients should only call this request if the corresponding capability
```
Its the DAP documentation for the `cancel` request, its mostly verbatim with small formatting tweaks. See https://microsoft.github.io/debug-adapter-protocol/specification#Base_Protocol_Cancel

ashgti committed:
Trying to isolate the Win32 timeout.

ashgti committed:
Fixing the EndOfFileError error code.

labath commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -671,8 +695,33 @@ void DAP::SetTarget(const lldb::SBTarget target) {
+      }
+    }
+
+    // Clear interrupt marker prior to handling the next request.
+    if (debugger.InterruptRequested())
+      debugger.CancelInterruptRequest();
```
You still have the same race here. The request could be cancelled between this line and when the active request field is set on line 702. (My complaint wasn't so much about the exact std::atomic type as it was about treating this flag as an atomic -- a mutex surrounding just a variable assignment is morally equivalent to an atomic). For this to work, I think the flag needs to be checked in the same critical section that handles  sets `m_active_request`. I believe it will be fine if you move this code to line 702, just before setting the active request.

ashgti replied:
Moved this up to 702

ashgti committed:
Adjusting includes and when we call CancelInterruptRequest.

ashgti committed:
Moving cancel checking into RequestHandler to try to conslidate the cancellation logic.

ashgti commented:
@JDevlieghere I moved some of the cancel checking logic into the BaseRequestHandler to try to consolidate things. There are some FIXME's around cleaning it up once all the requests have moved off the LegacyRequestHandler.

ashgti commented:
Anyone have any other comments on supporting 'cancel'?

ashgti committed:
Addressing feedback and reseting debugger interrupt requested sooner.

ashgti commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -671,11 +693,25 @@ void DAP::SetTarget(const lldb::SBTarget target) {
+
+      // Clear interrupt marker prior to handling the next request.
+      if (debugger.InterruptRequested())
+        debugger.CancelInterruptRequest();
```
Done, we still have this check, but it should be cleared sooner in `Send` as well.

ashgti commented on lldb/tools/lldb-dap/DAP.h:
```diff
@@ -397,6 +402,23 @@ struct DAP {
+  std::condition_variable m_queue_cv;
+
+  std::mutex m_cancelled_requests_mutex;
+  std::set<int64_t> m_cancelled_requests;
```
Done.

ashgti commented on lldb/tools/lldb-dap/Transport.h:
```diff
@@ -19,10 +19,39 @@
+  EndOfFileError() = default;
+
+  void log(llvm::raw_ostream &OS) const override {
+    OS << "End of file reached.";
```
Done.

ashgti commented on lldb/tools/lldb-dap/Transport.h:
```diff
@@ -19,10 +19,39 @@
+
+  void log(llvm::raw_ostream &OS) const override {
+    OS << "Operation timed out.";
+  }
```
Done.

ashgti committed:
Merge branch 'main' into cancel-request

ashgti committed:
Fixing post-merge changes.

ashgti commented:
Wanted to also double check with @labath as well, any additional thoughts?

labath commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -778,28 +817,121 @@ llvm::Error DAP::Disconnect(bool terminateDebuggee) {
+}
+
+void DAP::ClearCancelRequest(const CancelArguments &args) {
+  std::lock_guard<std::mutex> cancalled_requests_lock(
```
```suggestion
  std::lock_guard<std::mutex> lock(
```

It matches the name you use above, and I think the function is short enough..

ashgti replied:
I noticed in a few different places the naming was 'guard' for the lock_guard variable. Swapped all my variables to 'guard' instead.

labath commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -778,28 +817,121 @@ llvm::Error DAP::Disconnect(bool terminateDebuggee) {
+
+    while (!disconnecting) {
+      llvm::Expected<Message> next = transport.Read(std::chrono::seconds(1));
+      bool timeout = false;
+      bool eof = false;
+      if (llvm::Error Err = llvm::handleErrors(
+              next.takeError(),
+              [&](const EndOfFileError &E) -> llvm::Error {
+                eof = true;
+                return llvm::Error::success();
+              },
+              [&](const TimeoutError &) -> llvm::Error {
+                timeout = true;
+                return llvm::Error::success();
+              }))
+        return Err;
+
+      if (eof)
+        break;
```
Something like this would probably look better
```suggestion
      if (next.errorIsA<EndOfFileError>) {
        consumeError(next.takeError());
        break;
      }
      // same for TimeoutError
      if (!next)
        return next.takeError();

```

ashgti replied:
Done.

labath commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -778,28 +817,121 @@ llvm::Error DAP::Disconnect(bool terminateDebuggee) {
+
 llvm::Error DAP::Loop() {
-  auto cleanup = llvm::make_scope_exit([this]() {
+  std::future<llvm::Error> queue_reader = std::async([&]() -> llvm::Error {
```
I think this should explicitly use the std::launch::async flag as it won't work in the "deferred" policy and "If more than one flag is set, it is implementation-defined which policy is selected."

ashgti replied:
Done.

labath reviewed (approved):
Didn't look at the code all too closely but it looks okay to me. I believe my earlier concerns about cancellation races are resolved now.

ashgti committed:
Applying reviewer suggestions and have the reader loop break when the "disconnect" request is detected.

ashgti committed:
Merge remote-tracking branch 'origin/main' into cancel-request

ashgti committed:
Fixing post-merge issues.

ashgti commented:
PR #135638 should fix the build failure.
